javascript/ql/src/semmle/javascript/frameworks/Redux.qll

```diff
@@ -0,0 +1,1284 @@
+     */
+    private class ReduceReducers extends DataFlow::CallNode, DelegatingReducer {
+      ReduceReducers() {
+        this = API::moduleImport("reduce-reducers").getACall() or
+        this =
+          API::moduleImport(["redux-actions", "redux-ts-utils"])
+              .getMember("reduceReducers")
+              .getACall()
+      }
+
+      override DataFlow::Node getAPlainHandlerArg() {
+        result = getAnArgument()
+        or
+        result = getArgument(0).getALocalSource().(DataFlow::ArrayCreationNode).getAnElement()
+      }
+    }
+
+    /**
+     * A call to `createReducer`, for example:
+     *
+     * ```js
+     * let reducer = createReducer(initialState, (builder) => {
+     *   builder
+     *     .addCase(actionType1, (state, action) => { ... })
+     *     .addCase(actionType2, (state, action) => { ... });
+     * });
+     * ```
+     */
+    private class CreateReducer extends API::CallNode, DelegatingReducer {
+      CreateReducer() {
+        this = API::moduleImport("@reduxjs/toolkit").getMember("createReducer").getACall()
+      }
+
+      private API::Node getABuilderRef() {
+        result = getParameter(1).getParameter(0)
+        or
+        result = getABuilderRef().getAMember().getReturn()
+      }
+
+      override DataFlow::Node getActionHandlerArg(DataFlow::Node actionType) {
+        exists(API::CallNode addCase |
+          addCase = getABuilderRef().getMember("addCase").getACall() and
+          actionType = addCase.getArgument(0) and
+          result = addCase.getArgument(1)
+        )
+      }
+    }
+
+    /**
+     * A reducer created by a call to `createSlice`. Note that `createSlice` creates both
+     * reducers and actions; this class models the reducers only.
+     *
+     * For example:
+     * ```js
+     * let slice = createSlice({
+     *   name: 'mySlice',
+     *   reducers: {
+     *     actionType1: (state, action) => { ... },
+     *     actionType2: (state, action) => { ... },
+     *   },
+     *   extraReducers: (builder) => {
+     *     builder.addCase('actionType3', (state, action) => { ... })
+     *   }
+     * });
+     * export default slice.reducer;
+     * ```
+     */
+    private class CreateSliceReducer extends DelegatingReducer {
+      API::CallNode call;
+
+      CreateSliceReducer() {
+        call = API::moduleImport("@reduxjs/toolkit").getMember("createSlice").getACall() and
+        this = call.getReturn().getMember("reducer").getAnImmediateUse()
+      }
+
+      private API::Node getABuilderRef() {
+        result = call.getParameter(0).getMember("extraReducers").getParameter(0)
+        or
+        result = getABuilderRef().getAMember().getReturn()
+      }
+
+      override DataFlow::Node getActionHandlerArg(DataFlow::Node actionType) {
+        exists(string name |
+          result = call.getParameter(0).getMember("reducers").getMember(name).getARhs() and
+          actionType = call.getReturn().getMember("actions").getMember(name).getAnImmediateUse()
```

Comment:
From the qldoc:
> The `actionType` node may refer an action creator or a string value corresponding to `action.type`.

This happens because action-creator libraries return functions whose `toString` method has been overridden to return the corresponding action type, so the function itself can be used as the property name.

----

Also, I expanded the qldoc example a bit to mention the returned `.actions` property, as this predicate doesn't do anything without it.